[Bug]: install.py fails in uv venvs without pip (sqlite-vec backend)

### Bug Description

When running the installer in a uv-created virtual environment, the sqlite-vec backend install step calls python -m pip install sqlite-vec. uv venv environments can omit pip, causing the installer to fail with No module named pip and abort the sqlite-vec backend installation. The installer already detects uv earlier, but this backend-install path still hard-requires pip.

### Steps to Reproduce

  1. uv venv .venv
  2. Confirm pip is missing: .venv/bin/python3 -c "import pip" (expected to fail)
  3. Run installer for sqlite-vec: .venv/bin/python3 install.py --storage-backend sqlite_vec
  4. Observe failure during sqlite-vec installation

### Expected Behavior

If pip is unavailable but uv is installed, the installer should fall back to installing packages via uv pip (targeting the current interpreter), e.g. uv pip install --python <venv python> sqlite-vec, instead of failing.

### Actual Behavior

[3c] Installing sqlite_vec storage backend
    -> Installing SQLite-vec...
python3: No module named pip
    [ERROR] Failed to install SQLite-vec: Command '[python3, -m, pip, install, sqlite-vec]' returned non-zero exit status 1.
    [ERROR] Failed to install sqlite_vec storage backend

### Storage Backend

sqlite-vec (local)

### Operating System

macOS

### Python Version

Python 3.12.x

### MCP Memory Service Version

MCP Memory Service, version 8.52.2

### Installation Method

Source (git clone)

### Interface Used

Other

### Configuration

```shell

```

### Relevant Log Output

```shell

```

### Additional Context

This blocks fresh installs for users using uv-managed virtual environments. A PR is ready to fix by using uv pip install --python <sys.executable> when python -m pip is unavailable (and adding a unit test for the fallback).

### Pre-submission Checklist

- [x] I've searched existing issues and this is not a duplicate
- [x] I'm using the latest version (or specified version above)
- [x] I've included all required environment information
- [x] I've redacted sensitive information (API keys, tokens)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,8 @@ For older releases, see [CHANGELOG-HISTORIC.md](./CHANGELOG-HISTORIC.md).
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Fixed
+- Installer now supports `uv` virtual environments that don’t include `pip` by falling back to `uv pip` (targeting the active interpreter).
 
 ## [8.54.0] - 2025-12-23
 
@@ -2024,4 +2026,3 @@ The fallback implementation remains available for experimentation, but MS-MARCO'
 - `670fb74` - Phase 1: Automated detection (check_dev_setup.py, pre-commit hook, CLAUDE.md)
 - `9537259` - Phase 2: Runtime warnings (server.py) + developer documentation
 - `a17bcc7` - Phase 3: Interactive onboarding (install.py) + CI/CD validation
-

--- a/install.py
+++ b/install.py
@@ -187,6 +187,57 @@ except ImportError:
 # Global variable to store the uv executable path
 UV_EXECUTABLE_PATH = None
 
+def _pip_available() -> bool:
+    """Return True if `python -m pip` is available in the current environment."""
+    try:
+        subprocess.check_call(
+            [sys.executable, "-m", "pip", "--version"],
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        return True
+    except (subprocess.SubprocessError, OSError):
+        return False
+
+def _uv_executable() -> str | None:
+    """Return uv executable path if available (prefer detected path)."""
+    return UV_EXECUTABLE_PATH or shutil.which("uv")
+
+def _install_python_packages(
+    packages: list[str],
+    *,
+    extra_args: list[str] | None = None,
+    silent: bool = False,
+    env: dict[str, str] | None = None,
+) -> None:
+    """Install packages using pip if present, otherwise fall back to `uv pip`.
+
+    Raises:
+        subprocess.SubprocessError: If the installation command fails
+        OSError: If the installation command cannot be executed
+        RuntimeError: If neither pip nor uv are available
+    """
+    if _pip_available():
+        cmd: list[str] = [sys.executable, "-m", "pip", "install"]
+    else:
+        uv_path = _uv_executable()
+        if not uv_path:
+            raise RuntimeError("Neither pip nor uv could be found. Cannot install packages.")
+        cmd = [uv_path, "pip", "install", "--python", sys.executable]
+
+    if extra_args:
+        cmd.extend(extra_args)
+    cmd.extend(packages)
+
+    kwargs: dict[str, object] = {}
+    if env is not None:
+        kwargs["env"] = env
+    if silent:
+        kwargs["stdout"] = subprocess.DEVNULL
+        kwargs["stderr"] = subprocess.DEVNULL
+
+    subprocess.check_call(cmd, **kwargs)
+
 def print_header(text):
     """Print a formatted header."""
     print("\n" + "=" * 80)
@@ -446,7 +497,7 @@ def check_dependencies():
             if in_venv:
                 print_warning("pip could not be detected, but you're in a virtual environment. "
                             "If you're using uv or another alternative package manager, this is normal. "
-                            "Continuing installation...")
+                            "Continuing installation (will use uv where needed)...")
                 pip_installed = True  # Proceed anyway
             else:
                 print_error("pip is not installed. Please install pip first.")
@@ -461,14 +512,13 @@ def check_dependencies():
         # If pip is available, use it to install setuptools
         if pip_installed:
             try:
-                subprocess.check_call([sys.executable, '-m', 'pip', 'install', 'setuptools'], 
-                                    stdout=subprocess.DEVNULL)
+                _install_python_packages(["setuptools"], silent=True)
                 print_success("setuptools installed successfully")
-            except subprocess.SubprocessError:
+            except (subprocess.SubprocessError, OSError, RuntimeError):
                 # Check if in virtual environment
                 in_venv = sys.prefix != sys.base_prefix
                 if in_venv:
-                    print_warning("Failed to install setuptools with pip. If you're using an alternative package manager "
+                    print_warning("Failed to install setuptools automatically. If you're using an alternative package manager "
                                 "like uv, please install setuptools manually using that tool (e.g., 'uv pip install setuptools').")
                 else:
                     print_error("Failed to install setuptools. Please install it manually.")
@@ -487,14 +537,13 @@ def check_dependencies():
         # If pip is available, use it to install wheel
         if pip_installed:
             try:
-                subprocess.check_call([sys.executable, '-m', 'pip', 'install', 'wheel'], 
-                                    stdout=subprocess.DEVNULL)
+                _install_python_packages(["wheel"], silent=True)
                 print_success("wheel installed successfully")
-            except subprocess.SubprocessError:
+            except (subprocess.SubprocessError, OSError, RuntimeError):
                 # Check if in virtual environment
                 in_venv = sys.prefix != sys.base_prefix
                 if in_venv:
-                    print_warning("Failed to install wheel with pip. If you're using an alternative package manager "
+                    print_warning("Failed to install wheel automatically. If you're using an alternative package manager "
                                 "like uv, please install wheel manually using that tool (e.g., 'uv pip install wheel').")
                 else:
                     print_error("Failed to install wheel. Please install it manually.")
@@ -973,10 +1022,10 @@ def install_storage_backend(backend, system_info):
         # Standard installation for older Python versions
         try:
             print_info("Installing SQLite-vec...")
-            subprocess.check_call([sys.executable, '-m', 'pip', 'install', 'sqlite-vec'])
+            _install_python_packages(["sqlite-vec"])
             print_success("SQLite-vec installed successfully")
             return True
-        except subprocess.SubprocessError as e:
+        except (subprocess.SubprocessError, OSError, RuntimeError) as e:
             print_error(f"Failed to install SQLite-vec: {e}")
             return False
             
@@ -1078,14 +1127,14 @@ def install_sqlite_vec_python313(system_info):
         # Strategy 1: Try with uv pip
         strategies.append({
             'name': 'uv pip install',
-            'cmd': [uv_path, 'pip', 'install', 'sqlite-vec'],
+            'cmd': [uv_path, 'pip', 'install', '--python', sys.executable, 'sqlite-vec'],
             'description': 'Installing with uv package manager'
         })
         
         # Strategy 2: Try with uv pip and no-binary flag
         strategies.append({
             'name': 'uv pip install (source build)',
-            'cmd': [uv_path, 'pip', 'install', '--no-binary', ':all:', 'sqlite-vec'],
+            'cmd': [uv_path, 'pip', 'install', '--python', sys.executable, '--no-binary', ':all:', 'sqlite-vec'],
             'description': 'Building from source with uv'
         })
     
@@ -1388,7 +1437,10 @@ def _install_with_onnx(installer_cmd, install_mode, chosen_backend, env, using_h
             print_info("Using Python 3.13+ on macOS Intel - using SQLite-vec + ONNX configuration")
         
         # Install without ML dependencies
-        cmd = installer_cmd + ['install', '--no-deps'] + install_mode + ['.']
+        cmd = installer_cmd + ['install']
+        if len(installer_cmd) >= 2 and Path(installer_cmd[0]).stem == "uv" and installer_cmd[1] == "pip":
+            cmd += ['--python', sys.executable]
+        cmd += ['--no-deps'] + install_mode + ['.']
         print_info(f"Running: {' '.join(cmd)}")
         subprocess.check_call(cmd, env=env)
         
@@ -1413,7 +1465,7 @@ def _install_with_onnx(installer_cmd, install_mode, chosen_backend, env, using_h
             dependencies.append("tokenizers==0.20.3")
         
         # Install dependencies
-        subprocess.check_call([sys.executable, '-m', 'pip', 'install'] + dependencies)
+        _install_python_packages(dependencies)
         
         # Configure ONNX runtime
         print_info("Configuring to use ONNX runtime for inference without PyTorch...")
@@ -1436,7 +1488,7 @@ def _install_with_onnx(installer_cmd, install_mode, chosen_backend, env, using_h
             print_info("The service will use ONNX runtime for inference instead")
         
         return True
-    except subprocess.SubprocessError as e:
+    except (subprocess.SubprocessError, OSError, RuntimeError) as e:
         print_error(f"Failed to install with ONNX approach: {e}")
         return False
 
@@ -1453,7 +1505,10 @@ def _install_standard(installer_cmd, install_mode, env):
         bool: True if installation succeeded, False otherwise
     """
     try:
-        cmd = installer_cmd + ['install'] + install_mode + ['.']
+        cmd = installer_cmd + ['install']
+        if len(installer_cmd) >= 2 and Path(installer_cmd[0]).stem == "uv" and installer_cmd[1] == "pip":
+            cmd += ['--python', sys.executable]
+        cmd += install_mode + ['.']
         print_info(f"Running: {' '.join(cmd)}")
         subprocess.check_call(cmd, env=env)
         print_success("MCP Memory Service installed successfully")

--- a/scripts/installation/install.py
+++ b/scripts/installation/install.py
@@ -13,6 +13,7 @@ import shutil
 from pathlib import Path
 from typing import Tuple, Dict, Any, Optional
 import re
+import importlib.util
 
 def is_python_version_at_least(major, minor):
     """Check if current Python version is at least the specified version.
@@ -192,7 +193,15 @@ def install_package_safe(package, success_msg=None, error_msg=None, fallback_in_
     Returns:
         bool: True if installation succeeded OR if fallback was applied (see warning messages)
     """
-    cmd = [sys.executable, '-m', 'pip', 'install', package]
+    pip_module_available = importlib.util.find_spec("pip") is not None
+    uv_path = shutil.which("uv")
+    if pip_module_available:
+        cmd = [sys.executable, '-m', 'pip', 'install', package]
+    elif uv_path:
+        # uv environments commonly omit pip; use uv to install into the current interpreter
+        cmd = [uv_path, 'pip', 'install', '--python', sys.executable, package]
+    else:
+        cmd = [sys.executable, '-m', 'pip', 'install', package]
     default_success = success_msg or f"{package} installed successfully"
     default_error = error_msg or f"Failed to install {package}"
 
@@ -348,7 +357,7 @@ def check_dependencies():
             if in_venv:
                 print_warning("pip could not be detected, but you're in a virtual environment. "
                             "If you're using uv or another alternative package manager, this is normal. "
-                            "Continuing installation...")
+                            "Continuing installation (will use uv where needed)...")
                 pip_installed = True  # Proceed anyway
             else:
                 print_error("pip is not installed. Please install pip first.")
@@ -1090,7 +1099,10 @@ def _setup_storage_and_gpu_environment(args, system_info, gpu_info, env):
             
             # First try to install without ML dependencies
             try:
-                cmd = installer_cmd + ['install', '--no-deps'] + install_mode + ['.']
+                cmd = installer_cmd + ['install']
+                if len(installer_cmd) >= 2 and Path(installer_cmd[0]).stem == "uv" and installer_cmd[1] == "pip":
+                    cmd += ['--python', sys.executable]
+                cmd += ['--no-deps'] + install_mode + ['.']
                 success, _ = run_command_safe(
                     cmd,
                     success_msg="Package installed with --no-deps successfully",
@@ -1114,7 +1126,14 @@ def _setup_storage_and_gpu_environment(args, system_info, gpu_info, env):
                     dependencies.append("chromadb==0.5.23")
                 
                 # Install dependencies
-                cmd = [sys.executable, '-m', 'pip', 'install'] + dependencies
+                pip_module_available = importlib.util.find_spec("pip") is not None
+                uv_path = shutil.which("uv")
+                if pip_module_available:
+                    cmd = [sys.executable, '-m', 'pip', 'install'] + dependencies
+                elif uv_path:
+                    cmd = [uv_path, 'pip', 'install', '--python', sys.executable] + dependencies
+                else:
+                    cmd = [sys.executable, '-m', 'pip', 'install'] + dependencies
                 success, _ = run_command_safe(
                     cmd,
                     success_msg="Core dependencies installed successfully",
@@ -1167,7 +1186,10 @@ def _setup_storage_and_gpu_environment(args, system_info, gpu_info, env):
             print_info("Installing lightweight version (no ML dependencies by default)")
             print_info("For full functionality, use --with-ml flag or install with: pip install mcp-memory-service[ml]")
 
-        cmd = installer_cmd + ['install'] + install_mode + install_target
+        cmd = installer_cmd + ['install']
+        if len(installer_cmd) >= 2 and Path(installer_cmd[0]).stem == "uv" and installer_cmd[1] == "pip":
+            cmd += ['--python', sys.executable]
+        cmd += install_mode + install_target
         success, _ = run_command_safe(
             cmd,
             success_msg="MCP Memory Service installed successfully",

--- a/scripts/installation/install.py
+++ b/scripts/installation/install.py
@@ -201,7 +201,8 @@ def install_package_safe(package, success_msg=None, error_msg=None, fallback_in_
         # uv environments commonly omit pip; use uv to install into the current interpreter
         cmd = [uv_path, 'pip', 'install', '--python', sys.executable, package]
     else:
-        cmd = [sys.executable, '-m', 'pip', 'install', package]
+        print_error("Neither pip nor uv detected. Cannot install packages.")
+        return False
     default_success = success_msg or f"{package} installed successfully"
     default_error = error_msg or f"Failed to install {package}"
 
@@ -1125,15 +1126,11 @@ def _setup_storage_and_gpu_environment(args, system_info, gpu_info, env):
                 if args.with_chromadb:
                     dependencies.append("chromadb==0.5.23")
                 
-                # Install dependencies
-                pip_module_available = importlib.util.find_spec("pip") is not None
-                uv_path = shutil.which("uv")
-                if pip_module_available:
-                    cmd = [sys.executable, '-m', 'pip', 'install'] + dependencies
-                elif uv_path:
-                    cmd = [uv_path, 'pip', 'install', '--python', sys.executable] + dependencies
-                else:
-                    cmd = [sys.executable, '-m', 'pip', 'install'] + dependencies
+                # Install dependencies using the same installer selection as the main install
+                cmd = installer_cmd + ['install']
+                if len(installer_cmd) >= 2 and Path(installer_cmd[0]).stem == "uv" and installer_cmd[1] == "pip":
+                    cmd += ['--python', sys.executable]
+                cmd += dependencies
                 success, _ = run_command_safe(
                     cmd,
                     success_msg="Core dependencies installed successfully",

--- a/tests/unit/test_uv_no_pip_installer_fallback.py
+++ b/tests/unit/test_uv_no_pip_installer_fallback.py
@@ -80,3 +80,13 @@ def test_scripts_installer_uses_uv_when_pip_module_missing(monkeypatch):
     assert "--python" in recorded["cmd"]
     assert sys.executable in recorded["cmd"]
     assert "sqlite-vec" in recorded["cmd"]
+
+
+def test_scripts_installer_returns_false_when_no_pip_and_no_uv(monkeypatch):
+    from scripts.installation import install as scripts_install
+
+    monkeypatch.setattr(scripts_install.importlib.util, "find_spec", lambda name: None)
+    monkeypatch.setattr(scripts_install.shutil, "which", lambda name: None)
+
+    ok = scripts_install.install_package_safe("sqlite-vec", fallback_in_venv=False)
+    assert ok is False

--- a/tests/unit/test_uv_no_pip_installer_fallback.py
+++ b/tests/unit/test_uv_no_pip_installer_fallback.py
@@ -1,0 +1,82 @@
+import importlib.util
+import sys
+from pathlib import Path
+
+
+def _load_install_py_module(monkeypatch, base_dir: Path):
+    repo_root = Path(__file__).resolve().parents[2]
+    # Avoid touching user-global app support directories during import.
+    monkeypatch.setenv("MCP_MEMORY_BASE_DIR", str(base_dir))
+    install_py = repo_root / "install.py"
+    spec = importlib.util.spec_from_file_location("mcp_memory_service_install_py", install_py)
+    assert spec and spec.loader
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+    return module
+
+
+def test_install_py_uses_pip_when_available(monkeypatch, tmp_path):
+    install_module = _load_install_py_module(monkeypatch, tmp_path)
+
+    recorded = {}
+
+    def fake_check_call(cmd, **kwargs):
+        recorded["cmd"] = cmd
+        recorded["kwargs"] = kwargs
+        return 0
+
+    monkeypatch.setattr(install_module, "_pip_available", lambda: True)
+    monkeypatch.setattr(install_module, "subprocess", install_module.subprocess)
+    monkeypatch.setattr(install_module.subprocess, "check_call", fake_check_call)
+
+    install_module._install_python_packages(["sqlite-vec"], extra_args=["--no-deps"], silent=True)
+
+    assert recorded["cmd"][:4] == [sys.executable, "-m", "pip", "install"]
+    assert "--no-deps" in recorded["cmd"]
+    assert "sqlite-vec" in recorded["cmd"]
+    assert "stdout" in recorded["kwargs"]
+    assert "stderr" in recorded["kwargs"]
+
+
+def test_install_py_uses_uv_when_pip_missing(monkeypatch, tmp_path):
+    install_module = _load_install_py_module(monkeypatch, tmp_path)
+
+    recorded = {}
+
+    def fake_check_call(cmd, **kwargs):
+        recorded["cmd"] = cmd
+        recorded["kwargs"] = kwargs
+        return 0
+
+    monkeypatch.setattr(install_module, "_pip_available", lambda: False)
+    monkeypatch.setattr(install_module, "_uv_executable", lambda: "/opt/homebrew/bin/uv")
+    monkeypatch.setattr(install_module.subprocess, "check_call", fake_check_call)
+
+    install_module._install_python_packages(["sqlite-vec"])
+
+    assert recorded["cmd"][:3] == ["/opt/homebrew/bin/uv", "pip", "install"]
+    assert "--python" in recorded["cmd"]
+    assert sys.executable in recorded["cmd"]
+    assert "sqlite-vec" in recorded["cmd"]
+
+
+def test_scripts_installer_uses_uv_when_pip_module_missing(monkeypatch):
+    from scripts.installation import install as scripts_install
+
+    recorded = {}
+
+    def fake_run_command_safe(cmd, **kwargs):
+        recorded["cmd"] = cmd
+        recorded["kwargs"] = kwargs
+        return True, None
+
+    monkeypatch.setattr(scripts_install.importlib.util, "find_spec", lambda name: None)
+    monkeypatch.setattr(scripts_install.shutil, "which", lambda name: "/opt/homebrew/bin/uv")
+    monkeypatch.setattr(scripts_install, "run_command_safe", fake_run_command_safe)
+
+    ok = scripts_install.install_package_safe("sqlite-vec", fallback_in_venv=False)
+    assert ok is True
+    assert recorded["cmd"][:3] == ["/opt/homebrew/bin/uv", "pip", "install"]
+    assert "--python" in recorded["cmd"]
+    assert sys.executable in recorded["cmd"]
+    assert "sqlite-vec" in recorded["cmd"]

